url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/example/panelexample/FlyweightUIComponent.java

```diff
@@ -0,0 +1,309 @@
+		this.add(center, BorderLayout.CENTER);
+		chooserTree.addTreeSelectionListener(this);
+		chooserTree.getSelectionModel().setSelectionMode(
+				TreeSelectionModel.SINGLE_TREE_SELECTION);
+		useModelLabel.addItemListener(this);
+	}
+	
+	 public void paint(Graphics g) {
+		 super.paint(g);
+	 }
+
+	/**
+	 * the editor is lightweight so you need to install the graph cell
+	 * properties before using it.
+	 * 
+	 * @param value
+	 */
+	private void installValue(BusinessObjectWrapper2 value, boolean isEditing) {
+		oldModel = value;
+		newModel = value;
+		labelField.setText(new String(newModel.getLabel()));
+		labelField.setEnabled(true);
+		useModelLabel.setSelected(false);
+		chooserTree.setSelectionPath(null);
+
+		if (newModel.getValue() != null)
+			chooserTree.setSelectionPath(new TreePath(
+					((DefaultTreeModel) chooserTree.getModel())
+							.getPathToRoot(newModel.getValue())));
+	}
+
+	protected void installAttributes(Object value, CellView view,
+			boolean isEditing, JGraph graph) {
+		if (editorInsideCell && view.getBounds() != null)
+			setPreferredSize(new Dimension((int) view.getBounds().getWidth(),
+					(int) view.getBounds().getHeight()));
+		if (value instanceof BusinessObjectWrapper2)
+			installValue((BusinessObjectWrapper2) value, true);
+		else {
+			BusinessObjectWrapper2 wrapper = new BusinessObjectWrapper2();
+			if (value instanceof DefaultMutableTreeNode)
+				wrapper.setValue((DefaultMutableTreeNode) value);
+			wrapper.setLabel(value.toString());
+			installValue(wrapper, true);
+		}
+		
+		this.remove(east);
+		this.remove(west);
+		this.remove(north);
+		this.remove(south);
+		
+		east = new JPanel(new GridBagLayout());
+		west = new JPanel(new GridBagLayout());
+		north = new JPanel(new GridBagLayout());
+		south = new JPanel(new GridBagLayout());
+		
+
+		GridBagConstraints c = new GridBagConstraints();
+		c.weighty = 1.0;
+		c.weightx = 1.0;
+        c.gridx = 0;
+        c.gridy = 0;
+		
+		GraphModel model = graph.getModel();
+		int childCount = model.getChildCount(view.getCell());
+		for (int i = 0; i < childCount; i++) {
+			Object child = model.getChild(view.getCell(), i);
+			if (model.isPort(child)) {
+				CellView portView = graph.getGraphLayoutCache().getMapping(
+						child, false);
+				if (portView != null) {
+					Point2D point = GraphConstants.getOffset(portView.getAllAttributes());
+					String label = (String) model.getValue(portView.getCell());
+					JComponent portComponent;
+					if (isEditing)
+						portComponent = new JTextField(label);
+					else
+						portComponent = new JLabel(label);
+					portComponent.setPreferredSize(new Dimension(30, 20));
+					if (point.getX() == 0) {//left port
+						if (point.getY() == 100) {
+							c.anchor = GridBagConstraints.FIRST_LINE_START;
+						}
+						else if (point.getY() == 500) {
+							c.anchor = GridBagConstraints.LINE_START;
+						}
+						else {
+							c.anchor = GridBagConstraints.LAST_LINE_START;
+						}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 190:198, 202:210.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.